File: src/cloudtwin/providers/aws/sqs/handlers.py
"""
SQS HTTP handlers (AWS JSON / application/x-amz-json-1.0 protocol).

All SQS operations arrive as POST / with:
  Content-Type: application/x-amz-json-1.0
  X-Amz-Target: AmazonSQS.{OperationName}
"""

from __future__ import annotations

from fastapi import Request, Response
from fastapi.responses import JSONResponse

from cloudtwin.core.errors import CloudTwinError, NotFoundError
from cloudtwin.providers.aws.protocols.json_protocol import JsonProtocolRouter
from cloudtwin.providers.aws.sqs.service import SqsService


def _error(code: str, message: str, status: int = 400) -> JSONResponse:
    return JSONResponse({"__type": code, "message": message}, status_code=status)


def register_sqs_handlers(router: JsonProtocolRouter, service: SqsService) -> None:
    """Register all SQS JSON-protocol action handlers into the shared router."""

    async def create_queue(request: Request, body: dict) -> Response:
        name = body.get("QueueName", "")
        if not name:
            return _error("InvalidParameterValue", "QueueName is required")
        try:
            url = await service.create_queue(name)
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse({"QueueUrl": url})

    async def list_queues(request: Request, body: dict) -> Response:
        prefix = body.get("QueueNamePrefix", "")
        try:
            urls = await service.list_queues(prefix=prefix)
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse({"QueueUrls": urls})

    async def get_queue_url(request: Request, body: dict) -> Response:
        name = body.get("QueueName", "")
        if not name:
            return _error("InvalidParameterValue", "QueueName is required")
        try:
            url = await service.get_queue_url(name)
        except NotFoundError:
            return _error(
                "AWS.SimpleQueueService.NonExistentQueue",
                f"The specified queue {name!r} does not exist.",
                404,
            )
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse({"QueueUrl": url})

    async def send_message(request: Request, body: dict) -> Response:
        queue_url = body.get("QueueUrl", "")
        message_body = body.get("MessageBody", "")
        if not queue_url or not message_body:
            return _error("InvalidParameterValue", "QueueUrl and MessageBody are required")
        try:
            result = await service.send_message(queue_url, message_body)
        except NotFoundError:
            return _error(
                "AWS.SimpleQueueService.NonExistentQueue",
                "The specified queue does not exist.",
                404,
            )
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse(result)

    async def receive_message(request: Request, body: dict) -> Response:
        queue_url = body.get("QueueUrl", "")
        max_number = int(body.get("MaxNumberOfMessages", 1))
        if not queue_url:
            return _error("InvalidParameterValue", "QueueUrl is required")
        try:
            messages = await service.receive_messages(queue_url, max_number)
        except NotFoundError:
            return _error(
                "AWS.SimpleQueueService.NonExistentQueue",
                "The specified queue does not exist.",
                404,
            )
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse({"Messages": messages})

    async def delete_message(request: Request, body: dict) -> Response:
        queue_url = body.get("QueueUrl", "")
        receipt_handle = body.get("ReceiptHandle", "")
        if not queue_url or not receipt_handle:
            return _error("InvalidParameterValue", "QueueUrl and ReceiptHandle are required")
        try:
            await service.delete_message(queue_url, receipt_handle)
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse({})

    async def delete_queue(request: Request, body: dict) -> Response:
        queue_url = body.get("QueueUrl", "")
        if not queue_url:
            return _error("InvalidParameterValue", "QueueUrl is required")
        try:
            await service.delete_queue(queue_url)
        except CloudTwinError as exc:
            return _error(exc.code, exc.message, exc.http_status)
        return JSONResponse({})

    router.register("AmazonSQS.CreateQueue", create_queue)
    router.register("AmazonSQS.ListQueues", list_queues)
    router.register("AmazonSQS.GetQueueUrl", get_queue_url)
    router.register("AmazonSQS.SendMessage", send_message)
    router.register("AmazonSQS.ReceiveMessage", receive_message)
    router.register("AmazonSQS.DeleteMessage", delete_message)
    router.register("AmazonSQS.DeleteQueue", delete_queue)

File: rust/cloudtwin-lite/src/aws/sqs/handlers.rs
//! SQS AWS JSON protocol handlers.
//!
//! This file is the thin edge between the AWS `X-Amz-Target` protocol and the
//! queue service methods.

use std::sync::Arc;

use axum::{
    http::StatusCode,
    response::{IntoResponse, Response},
};

use super::service::SqsService;
use crate::proto::json_error_response;
use crate::AppState;

fn svc(state: &Arc<AppState>) -> SqsService {
    SqsService::new(state.db.clone(), state.cfg.base_url.clone())
}

pub async fn handle_json(state: &Arc<AppState>, target: &str, body: serde_json::Value) -> Response {
    match target {
        "AmazonSQS.CreateQueue" => {
            let name = body.get("QueueName").and_then(|v| v.as_str()).unwrap_or("");
            if name.is_empty() {
                return json_error_response(
                    StatusCode::BAD_REQUEST,
                    "MissingParameter",
                    "QueueName required",
                );
            }
            match svc(state).create_queue(name).await {
                Ok(url) => (
                    StatusCode::OK,
                    axum::Json(serde_json::json!({ "QueueUrl": url })),
                )
                    .into_response(),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.ListQueues" => {
            let prefix = body
                .get("QueueNamePrefix")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            match svc(state).list_queues(prefix).await {
                Ok(urls) => (
                    StatusCode::OK,
                    axum::Json(serde_json::json!({ "QueueUrls": urls })),
                )
                    .into_response(),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.GetQueueUrl" => {
            let name = body.get("QueueName").and_then(|v| v.as_str()).unwrap_or("");
            match svc(state).get_queue_url(name).await {
                Ok(url) => (
                    StatusCode::OK,
                    axum::Json(serde_json::json!({ "QueueUrl": url })),
                )
                    .into_response(),
                Err(e) if e.to_string().contains("QueueDoesNotExist") => json_error_response(
                    StatusCode::BAD_REQUEST,
                    "QueueDoesNotExist",
                    &e.to_string(),
                ),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.SendMessage" => {
            let url = body.get("QueueUrl").and_then(|v| v.as_str()).unwrap_or("");
            let body_str = body
                .get("MessageBody")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            match svc(state).send_message(url, body_str).await {
                Ok((mid, _)) => (
                    StatusCode::OK,
                    axum::Json(serde_json::json!({
                        "MessageId":        mid,
                        "MD5OfMessageBody": SqsService::md5_body(body_str),
                    })),
                )
                    .into_response(),
                Err(e) if e.to_string().contains("QueueDoesNotExist") => json_error_response(
                    StatusCode::BAD_REQUEST,
                    "QueueDoesNotExist",
                    &e.to_string(),
                ),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.ReceiveMessage" => {
            let url = body.get("QueueUrl").and_then(|v| v.as_str()).unwrap_or("");
            let max = body
                .get("MaxNumberOfMessages")
                .and_then(|v| v.as_i64())
                .unwrap_or(1);
            match svc(state).receive_messages(url, max).await {
                Ok(msgs) => {
                    let items: Vec<_> = msgs
                        .iter()
                        .map(|m| {
                            serde_json::json!({
                                "MessageId":     m.message_id,
                                "ReceiptHandle": m.receipt_handle,
                                "Body":          m.body,
                                "MD5OfBody":     SqsService::md5_body(&m.body),
                            })
                        })
                        .collect();
                    (
                        StatusCode::OK,
                        axum::Json(serde_json::json!({ "Messages": items })),
                    )
                        .into_response()
                }
                Err(e) if e.to_string().contains("QueueDoesNotExist") => json_error_response(
                    StatusCode::BAD_REQUEST,
                    "QueueDoesNotExist",
                    &e.to_string(),
                ),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.DeleteMessage" => {
            let rh = body
                .get("ReceiptHandle")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            match svc(state).delete_message(rh).await {
                Ok(_) => (StatusCode::OK, axum::Json(serde_json::json!({}))).into_response(),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.DeleteQueue" => {
            let url = body.get("QueueUrl").and_then(|v| v.as_str()).unwrap_or("");
            match svc(state).delete_queue(url).await {
                Ok(_) => (StatusCode::OK, axum::Json(serde_json::json!({}))).into_response(),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.ChangeMessageVisibility" => {
            let _url = body.get("QueueUrl").and_then(|v| v.as_str()).unwrap_or("");
            let rh = body
                .get("ReceiptHandle")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let to = body
                .get("VisibilityTimeout")
                .and_then(|v| v.as_i64())
                .unwrap_or(30);
            match svc(state).change_visibility(rh, to).await {
                Ok(_) => (StatusCode::OK, axum::Json(serde_json::json!({}))).into_response(),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        "AmazonSQS.GetQueueAttributes" => {
            let url = body.get("QueueUrl").and_then(|v| v.as_str()).unwrap_or("");
            match svc(state).get_queue_attributes(url).await {
                Ok(attrs) => (StatusCode::OK, axum::Json(attrs)).into_response(),
                Err(e) if e.to_string().contains("QueueDoesNotExist") => json_error_response(
                    StatusCode::BAD_REQUEST,
                    "QueueDoesNotExist",
                    &e.to_string(),
                ),
                Err(e) => json_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "InternalError",
                    &e.to_string(),
                ),
            }
        }
        _ => json_error_response(
            StatusCode::BAD_REQUEST,
            "InvalidAction",
            &format!("Unknown SQS target: {target}"),
        ),
    }
}
